File: src/shared/media/compose-layout.ts
/**
 * Pure layout maths for the composer — no DOM, no canvas. Given an output size
 * and a layout choice, where do the video and map panes go, and how does each
 * source map into its pane (object-fit cover/contain → drawImage rects).
 *
 * Kept pure so the same numbers drive the live preview and (later) the export,
 * and so the fiddly geometry is unit-tested.
 */

export interface Rect {
  x: number;
  y: number;
  w: number;
  h: number;
}

export type LayoutKind = 'side-by-side' | 'stacked' | 'pip-map' | 'pip-video';
export type Corner = 'tl' | 'tr' | 'bl' | 'br';
export type Fit = 'cover' | 'contain';

export interface PaneRects {
  video: Rect;
  map: Rect;
}

/** Round to the nearest even integer (encoders prefer even dimensions). */
export function even(n: number): number {
  return 2 * Math.round(n / 2);
}

/** Output pixel size from an aspect ratio and a target long edge. */
export function outputSize(aspectW: number, aspectH: number, longEdge: number): {
  w: number;
  h: number;
} {
  if (aspectW >= aspectH) {
    return { w: even(longEdge), h: even((longEdge * aspectH) / aspectW) };
  }
  return { w: even((longEdge * aspectW) / aspectH), h: even(longEdge) };
}

function cornerRect(
  corner: Corner,
  outW: number,
  outH: number,
  w: number,
  h: number,
  margin: number,
): Rect {
  const x = corner === 'tl' || corner === 'bl' ? margin : outW - w - margin;
  const y = corner === 'tl' || corner === 'tr' ? margin : outH - h - margin;
  return { x, y, w, h };
}

/**
 * The video and map pane rectangles within an `outW`×`outH` frame.
 * - side-by-side / stacked: `split` (0.1..0.9) is the video's fraction.
 * - pip-map / pip-video: the inset pane is `inset` (0.15..0.6) of the width, in
 *   `corner`, over the full-frame other pane.
 */
export function paneRects(
  outW: number,
  outH: number,
  layout: LayoutKind,
  split: number,
  inset: number,
  corner: Corner,
): PaneRects {
  const s = Math.min(0.9, Math.max(0.1, split));
  if (layout === 'side-by-side') {
    const vw = Math.round(outW * s);
    return {
      video: { x: 0, y: 0, w: vw, h: outH },
      map: { x: vw, y: 0, w: outW - vw, h: outH },
    };
  }
  if (layout === 'stacked') {
    const vh = Math.round(outH * s);
    return {
      video: { x: 0, y: 0, w: outW, h: vh },
      map: { x: 0, y: vh, w: outW, h: outH - vh },
    };
  }
  const full: Rect = { x: 0, y: 0, w: outW, h: outH };
  const frac = Math.min(0.6, Math.max(0.15, inset));
  const iw = Math.round(outW * frac);
  const ih = Math.round((iw * outH) / outW); // inset keeps the output aspect
  const margin = Math.round(outW * 0.03);
  const pip = cornerRect(corner, outW, outH, iw, ih, margin);
  return layout === 'pip-map'
    ? { video: full, map: pip }
    : { video: pip, map: full };
}

export interface DrawRect {
  sx: number;
  sy: number;
  sw: number;
  sh: number;
  dx: number;
  dy: number;
  dw: number;
  dh: number;
}

/**
 * Map a `srcW`×`srcH` source into `dst` with object-fit semantics:
 * - cover: crop the source to fill the pane (no bars).
 * - contain: fit the whole source inside the pane (letterboxed).
 * Returns the source-crop and destination rects for a single `drawImage`.
 */
export function fitRect(srcW: number, srcH: number, dst: Rect, mode: Fit): DrawRect {
  if (srcW <= 0 || srcH <= 0) {
    return { sx: 0, sy: 0, sw: srcW, sh: srcH, dx: dst.x, dy: dst.y, dw: dst.w, dh: dst.h };
  }
  const srcAR = srcW / srcH;
  const dstAR = dst.w / dst.h;
  if (mode === 'cover') {
    let sw = srcW;
    let sh = srcH;
    if (srcAR > dstAR) sw = srcH * dstAR;
    else sh = srcW / dstAR;
    return {
      sx: (srcW - sw) / 2,
      sy: (srcH - sh) / 2,
      sw,
      sh,
      dx: dst.x,
      dy: dst.y,
      dw: dst.w,
      dh: dst.h,
    };
  }
  let dw = dst.w;
  let dh = dst.h;
  if (srcAR > dstAR) dh = dst.w / srcAR;
  else dw = dst.h * srcAR;
  return {
    sx: 0,
    sy: 0,
    sw: srcW,
    sh: srcH,
    dx: dst.x + (dst.w - dw) / 2,
    dy: dst.y + (dst.h - dh) / 2,
    dw,
    dh,
  };
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Media/ComposeLayout.swift
// Pure layout maths for the composer — given an output size and a layout
// choice, where do the video and map panes go, and how does each source map
// into its pane (object-fit cover/contain → drawImage rects). Port of
// `src/shared/media/compose-layout.ts`.
//
// The web's `Rect { x, y, w, h }` is the kernel's `Rect` (`Geometry.swift`,
// `width`/`height`), its `Fit` the one `Framing.swift` already declares, and
// `Corner` is declared here because `MediaLayout.swift` reads it from this
// module exactly as the web does.

import Foundation

public enum LayoutKind: String, Codable, Sendable {
    case sideBySide = "side-by-side"
    case stacked
    case pipMap = "pip-map"
    case pipVideo = "pip-video"
}

public enum Corner: String, Codable, Sendable {
    case tl, tr, bl, br
}

public struct PaneRects: Equatable, Sendable {
    public var video: Rect
    public var map: Rect
    public init(video: Rect, map: Rect) { self.video = video; self.map = map }
}

/// JavaScript's `Math.round`: halves go toward +∞, never away from zero.
private func roundHalfUp(_ v: Double) -> Double {
    let down = v.rounded(.down)
    return v - down >= 0.5 ? down + 1 : down
}

/// Round to the nearest even integer (encoders prefer even dimensions).
public func even(_ n: Double) -> Double {
    2 * roundHalfUp(n / 2)
}

/// Output pixel size from an aspect ratio and a target long edge.
public func outputSize(_ aspectW: Double, _ aspectH: Double, _ longEdge: Double) -> Size {
    if aspectW >= aspectH {
        return Size(even(longEdge), even((longEdge * aspectH) / aspectW))
    }
    return Size(even((longEdge * aspectW) / aspectH), even(longEdge))
}

private func cornerRect(_ corner: Corner, _ outW: Double, _ outH: Double, _ w: Double, _ h: Double, _ margin: Double) -> Rect {
    let x = corner == .tl || corner == .bl ? margin : outW - w - margin
    let y = corner == .tl || corner == .tr ? margin : outH - h - margin
    return Rect(x, y, w, h)
}

/// The video and map pane rectangles within an `outW`×`outH` frame.
/// - side-by-side / stacked: `split` (0.1..0.9) is the video's fraction.
/// - pip-map / pip-video: the inset pane is `inset` (0.15..0.6) of the width, in
///   `corner`, over the full-frame other pane.
public func paneRects(_ outW: Double, _ outH: Double, _ layout: LayoutKind, _ split: Double, _ inset: Double, _ corner: Corner) -> PaneRects {
    let s = min(0.9, max(0.1, split))
    if layout == .sideBySide {
        let vw = roundHalfUp(outW * s)
        return PaneRects(video: Rect(0, 0, vw, outH), map: Rect(vw, 0, outW - vw, outH))
    }
    if layout == .stacked {
        let vh = roundHalfUp(outH * s)
        return PaneRects(video: Rect(0, 0, outW, vh), map: Rect(0, vh, outW, outH - vh))
    }
    let full = Rect(0, 0, outW, outH)
    let frac = min(0.6, max(0.15, inset))
    let iw = roundHalfUp(outW * frac)
    let ih = roundHalfUp((iw * outH) / outW) // inset keeps the output aspect
    let margin = roundHalfUp(outW * 0.03)
    let pip = cornerRect(corner, outW, outH, iw, ih, margin)
    return layout == .pipMap ? PaneRects(video: full, map: pip) : PaneRects(video: pip, map: full)
}

/// The source crop and the destination rect of one `drawImage`.
public struct DrawRect: Equatable, Sendable {
    public var sx: Double
    public var sy: Double
    public var sw: Double
    public var sh: Double
    public var dx: Double
    public var dy: Double
    public var dw: Double
    public var dh: Double

    public init(sx: Double, sy: Double, sw: Double, sh: Double, dx: Double, dy: Double, dw: Double, dh: Double) {
        self.sx = sx; self.sy = sy; self.sw = sw; self.sh = sh
        self.dx = dx; self.dy = dy; self.dw = dw; self.dh = dh
    }
}

/// Map a `srcW`×`srcH` source into `dst` with object-fit semantics:
/// - cover: crop the source to fill the pane (no bars).
/// - contain: fit the whole source inside the pane (letterboxed).
public func fitRect(_ srcW: Double, _ srcH: Double, _ dst: Rect, _ mode: Fit) -> DrawRect {
    if srcW <= 0 || srcH <= 0 {
        return DrawRect(sx: 0, sy: 0, sw: srcW, sh: srcH, dx: dst.x, dy: dst.y, dw: dst.width, dh: dst.height)
    }
    let srcAR = srcW / srcH
    let dstAR = dst.width / dst.height
    if mode == .cover {
        var sw = srcW
        var sh = srcH
        if srcAR > dstAR { sw = srcH * dstAR } else { sh = srcW / dstAR }
        return DrawRect(sx: (srcW - sw) / 2, sy: (srcH - sh) / 2, sw: sw, sh: sh,
                        dx: dst.x, dy: dst.y, dw: dst.width, dh: dst.height)
    }
    var dw = dst.width
    var dh = dst.height
    if srcAR > dstAR { dh = dst.width / srcAR } else { dw = dst.height * srcAR }
    return DrawRect(sx: 0, sy: 0, sw: srcW, sh: srcH,
                    dx: dst.x + (dst.width - dw) / 2, dy: dst.y + (dst.height - dh) / 2, dw: dw, dh: dh)
}
